File: server/Rendalicce/Domain/Users/PasswordValidator.cs
using FluentValidation;

namespace Rendalicce.Domain.Users;

public class PasswordValidator : AbstractValidator<string>
{
    public PasswordValidator()
    {
        // Rule for minimum 8 characters
        RuleFor(password => password)
            .NotEmpty()
            .MinimumLength(8).WithMessage("Lozinka mora imati minimalno 8 znakova.");

        // Rule for at least one digit
        RuleFor(password => password)
            .Matches(@"\d").WithMessage("Lozinka mora sadržavati najmanje jednu znamenku.");

        // Rule for at least one uppercase letter
        RuleFor(password => password)
            .Matches(@"[A-Z]").WithMessage("Lozinka mora sadržavati veliko slovo.");

        // Rule for at least one lowercase letter
        RuleFor(password => password)
            .Matches(@"[a-z]").WithMessage("Lozinka mora sadržavati malo slovo.");

        // Rule for at least one special character
        RuleFor(password => password)
            .Matches(@"[\W_]").WithMessage("Lozinka mora sadržavati poseban znak.");
    }
}

File: client/src/app/pages/authentication/register/password-validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

export function passwordValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    const value = control.value;
    if (!value) {
      return { required: 'Password is required.' }; // Empty password
    }

    const errors: ValidationErrors = {};

    // Rule for minimum length of 8 characters
    if (value.length < 8) {
      errors['minLength'] = 'Password must be at least 8 characters long.';
    }

    // Rule for at least one digit
    if (!/\d/.test(value)) {
      errors['digit'] = 'Password must contain at least one digit.';
    }

    // Rule for at least one uppercase letter
    if (!/[A-Z]/.test(value)) {
      errors['uppercase'] = 'Password must contain at least one uppercase letter.';
    }

    // Rule for at least one lowercase letter
    if (!/[a-z]/.test(value)) {
      errors['lowercase'] = 'Password must contain at least one lowercase letter.';
    }

    // Rule for at least one special character
    if (!/[\W_]/.test(value)) {
      errors['specialCharacter'] = 'Password must contain at least one special character.';
    }

    // Return errors if any, or null if the password is valid
    return Object.keys(errors).length > 0 ? errors : null;
  };
}
